### PR TITLE
[Evaluation] [Refactoring] Worker-wrapper-transformed the CEK machine

### PR DESCRIPTION
**Pull request by effectfully**
This seems to make the machine 1-2% faster. I was doing benchmarks yesterday all day long and it seems that tracking 1-2% optimizations is not something I can do on my machine even when turning off the browsers, messengers, the IDE etc. This refactoring results in minimal changes in Core, which makes me somewhat confident that it does improves over the existing state of the CEK machine. @kwxm could you run the benchmarks on your machine and tell if there's any noticeable difference?

### CODE REVIEW

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs [hunk 1]
```diff
@@ -93,6 +93,20 @@ Now clearly 'computeCek' cannot be inlined in 'runCek' whether it's exported or
...
 though the rest of the user-facing API (which 'runCek' is a part of) is defined downstream.
 
+Another problem is handling mutual recursion in the 'computeCek'/'returnCek'/'forceEvaluate'/etc
+family. If we keep it on the top level, GHC won't be able to pull the @Ix fun@ constraint out of
+the family (confirmed by inspecting Core: GHC thinks that since the superclass constraints
+populating the dictionary representing the @Ix fun@ constraint are redundant, they can be replaced
+with calls to 'error' in a recursive call, but that changes the dictionary and so it can no longer
+be pulled out of recursion). But that entails passing a redundant argument around, which slows down
+the machine a tiny little bit. Hence we perform the worker-wrapper transformation manually.
+However that allows GHC to inline almost all of the machine into a single definition (with a bunch
```
Comment (michaelpj) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:103]: :see_no_evil:

Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:103]: I would have hoped that would be faster!

**Review by michaelpj — APPROVED** [review]
Seems fine, although it's a bit sad. Maybe worth filing a GHC bug for the issue where changing the dictionary prevents the argument being hoisted?

**Comment by bezirg** [comment]
wow, a `NOINLINE` that improves speed!

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs (hunk 1 above)
Comment (effectfully) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:103]: Yes, but I tried it thrice and it was slower every time. But it seems whether my laptop is charging or not has an impact, so maybe I should try one more time (but I'm a bit sick of benchmarking).

**Comment by effectfully** [comment]
> wow, a `NOINLINE` that improves speed!

Allegedly.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs:103]: Nah, it's fine for now.

**Comment by kwxm** [comment]
> @kwxm could you run the benchmarks on your machine and tell if there's any noticeable difference?

There doesn't seem to be any significant difference between the validation benchmarks ...

```
Contract                 master          new
---------------------------------------------------------
crowdfunding/1           5.121 ms	5.169 ms    +0.9%
crowdfunding/2           5.262 ms	5.253 ms    -0.2%
crowdfunding/3           5.281 ms	5.223 ms    -1.1%
crowdfunding/4           3.250 ms	3.256 ms    +0.2%
crowdfunding/5           3.248 ms	3.233 ms    -0.5%
future/1                 3.124 ms	3.169 ms    +1.4%
future/2                 6.022 ms	6.039 ms    +0.3%
future/3                 6.028 ms	6.058 ms    +0.5%
future/4                 8.376 ms	8.250 ms    -1.5%
future/5                 9.754 ms	9.735 ms    -0.2%
future/6                 8.387 ms	8.434 ms    +0.6%
future/7                 9.717 ms	9.725 ms    +0.1%
multisigSM/1             6.181 ms	6.226 ms    +0.7%
multisigSM/2             6.224 ms	6.289 ms    +1.0%
multisigSM/3             6.265 ms	6.267 ms    +0.0%
multisigSM/4             6.394 ms	6.412 ms    +0.3%
multisigSM/5             7.321 ms	7.338 ms    +0.2%
multisigSM/6             6.161 ms	6.209 ms    +0.8%
multisigSM/7             6.247 ms	6.206 ms    -0.7%
multisigSM/8             6.325 ms	6.325 ms     0.0%
multisigSM/9             6.422 ms	6.359 ms    -1.0%
multisigSM/10            7.408 ms	7.334 ms    -1.0%
vesting/1                5.308 ms	5.222 ms    -1.6%
vesting/2                4.702 ms	4.708 ms    +0.1%
vesting/3                5.294 ms	5.257 ms    -0.7%
marlowe/trustfund/1      17.13 ms	17.11 ms    -0.1%
marlowe/trustfund/2      14.40 ms	13.90 ms    -3.5%
marlowe/zerocoupon/1     17.24 ms	17.31 ms    +0.4%
marlowe/zerocoupon/2     12.66 ms	12.84 ms    +1.4%
```

... but things seem to have got a tiny bit _worse_ with the nofib ones:

```
Benchmark                master         new
---------------------------------------------------------
clausify/formula1        230.8 ms	242.0 ms    +4.9%
clausify/formula2        298.4 ms	301.4 ms    +1.0%
clausify/formula3        820.5 ms	831.9 ms    +1.4%
clausify/formula4        1.124 s	1.133 s     +0.8%
clausify/formula5        5.377 s	5.568 s     +3.6%
knights/4x4              587.5 ms	595.1 ms    +1.3%
knights/6x6              1.722 s	1.740 s     +1.0%
knights/8x8              2.928 s	2.963 s     +1.2%
primetest/05digits       369.8 ms	373.0 ms    +0.9%
primetest/08digits       676.6 ms	687.6 ms    +1.6%
primetest/10digits       947.1 ms	955.6 ms    +0.9%
primetest/20digits       1.902 s	1.906 s     +0.2%
primetest/30digits       2.781 s	2.761 s     -0.7%
primetest/40digits       3.683 s	3.697 s     +0.4%
primetest/50digits       3.627 s	3.648 s     +0.6%
queens4x4/bt             99.98 ms	100.1 ms    +0.1%
queens4x4/bm             129.9 ms	129.8 ms    -0.1%
queens4x4/bjbt1          123.1 ms	123.4 ms    +0.2%
queens4x4/bjbt2          131.8 ms	132.1 ms    +0.2%
queens4x4/fc             313.0 ms	314.0 ms    +0.3%
queens5x5/bt             1.341 s	1.347 s     +0.4%
queens5x5/bm             1.469 s	1.466 s     -0.2%
queens5x5/bjbt1          1.573 s	1.585 s     +0.8%
queens5x5/bjbt2          1.656 s	1.663 s     +0.4%
queens5x5/fc             3.958 s	3.987 s     +0.7%
```

I pulled the master branch just before doing this.  I'm not sure if anything else has changed in master since your branch.

**Review by kwxm — APPROVED** [review]
I don't have a lot to say since I don't have a great understanding of what GHC's doing internally. Wrapping everything up inside a single function does seem at least aesthetically sensible, since we're never going to call, eg, `returnCek` anywhere else.  What's the general feeling amongst Haskell experts about doing these sort of manual transformations to get the compiler to behave nicely?  I can imagine carefully crafting your code to make it really efficient but then ending up back at square one when they decide to introduce some clever new optimisation in GHC.

I was looking at this code in connection with costing a few days ago and I wonder if it would make any difference to make the lambda/builtin cases of `forceEvaluate` and `applyEvaluate` directly into cases within `computeCek`.  I doubt if that would make any difference to performance, but it might make the code a little more comprehensible.  I was going to suggest doing the same for `applyBuiltin` but annoyingly that can be called in two different places (because of the possibility of having a force at the very end of a builtin), so we probably do want to make it into a separate function.

Another point is that we have to look up builtins _twice_, once when we encounter `Builtin` in `computeCek` and once when we're about to call `applyBuiltin`.  Would it be helpful to stash everything in the `VBuiltin` the first time?  We'd be trading a monadic lookup for an even bigger `VBuiltin`, so it might not gain us anything.

**Comment by effectfully** [comment]
> ... but things seem to have got a tiny bit worse with the nofib ones:

Well, then I'm converting this to a draft. Could also post the direct benchmarking results, not just the comparison? I wonder if there's any variance.

I'll also run `nofib` to replicate the results.

Thank you!

**Comment by kwxm** [comment]
> Well, then I'm converting this to a draft. Could you also post the direct benchmarking results, not just the comparison? I wonder if there's any variance.

Here you are.

[nofib-before-ww.txt](https://github.com/input-output-hk/plutus/files/6203778/nofib-before-ww.txt)
[nofib-after-ww.txt](https://github.com/input-output-hk/plutus/files/6203779/nofib-after-ww.txt)
[validation-before-ww.txt](https://github.com/input-output-hk/plutus/files/6203781/validation-before-ww.txt)
[validation-after-ww.txt](https://github.com/input-output-hk/plutus/files/6203782/validation-after-ww.txt)

**Comment by effectfully** [comment]
So I booted without X, killed all processes that I thought could impact benchmarks, prebuild the benchmarks (and let my laptop to cool down) before running them, made sure the laptop was not charging, added the `-L15` option that @kwxm has dug out and got the following impressively consistent results for the `validation` benchmarks:

[results-clean-no-ww-15.txt](https://github.com/input-output-hk/plutus/files/6223532/results-clean-no-ww-15.txt)
[results-clean-ww-15.txt](https://github.com/input-output-hk/plutus/files/6223542/results-clean-ww-15.txt)

For most benchmarks evaluation times are within +- 0.006ms (all benchmarks run for several ms), although Marlowe benchmarks behave slightly worse (still little variation though).

And now we could safely conclude that this PR did make things worse:

```
crowdfunding/1           3.142 ms	3.171 ms    +0.9%
crowdfunding/2           3.175 ms	3.207 ms    +1.0%
crowdfunding/3           3.175 ms	3.208 ms    +1.0%
crowdfunding/4           1.942 ms	1.957 ms    +0.8%
crowdfunding/5           1.943 ms	1.956 ms    +0.7%
future/1                 1.914 ms	1.928 ms    +0.7%
future/2                 3.660 ms	3.699 ms    +1.1%
future/3                 3.661 ms	3.696 ms    +1.0%
future/4                 4.886 ms	4.936 ms    +1.0%
future/5                 5.743 ms	5.815 ms    +1.3%
future/6                 4.890 ms	4.934 ms    +0.9%
future/7                 5.745 ms	5.815 ms    +1.2%
multisigSM/1             3.613 ms	3.652 ms    +1.1%
multisigSM/2             3.619 ms	3.664 ms    +1.2%
multisigSM/3             3.658 ms	3.699 ms    +1.1%
multisigSM/4             3.693 ms	3.733 ms    +1.1%
multisigSM/5             4.300 ms	4.339 ms    +0.9%
multisigSM/6             3.607 ms	3.651 ms    +1.2%
multisigSM/7             3.617 ms	3.658 ms    +1.1%
multisigSM/8             3.659 ms	3.701 ms    +1.1%
multisigSM/9             3.697 ms	3.735 ms    +1.0%
multisigSM/10            4.304 ms	4.345 ms    +1.0%
vesting/1                3.162 ms	3.193 ms    +1.0%
vesting/2                2.806 ms	2.841 ms    +1.2%
vesting/3                3.162 ms	3.190 ms    +0.9%
marlowe/trustfund/1      9.995 ms	10.06 ms    +0.7%
marlowe/trustfund/2      7.909 ms	7.981 ms    +0.9%
marlowe/zerocoupon/1     10.15 ms	10.24 ms    +0.9%
marlowe/zerocoupon/2     7.236 ms	7.287 ms    +0.7%
```

except I also compared the previous commit in this branch with the current `master` (albeit without `-L15`, but the results are still fairly consistent) and got:

```
crowdfunding/1           3.142 ms	3.151 ms    +0.3%
crowdfunding/2           3.175 ms	3.191 ms    +0.5%
crowdfunding/3           3.175 ms	3.195 ms    +0.6%
crowdfunding/4           1.942 ms	1.951 ms    +0.5%
crowdfunding/5           1.943 ms	1.953 ms    +0.5%
future/1                 1.914 ms	1.918 ms    +0.2%
future/2                 3.660 ms	3.679 ms    +0.5%
future/3                 3.661 ms	3.675 ms    +0.4%
future/4                 4.886 ms	4.933 ms    +1.0%
future/5                 5.743 ms	5.788 ms    +0.8%
future/6                 4.890 ms	4.949 ms    +1.2%
future/7                 5.745 ms	5.777 ms    +0.6%
multisigSM/1             3.613 ms	3.652 ms    +1.1%
multisigSM/2             3.619 ms	3.661 ms    +1.2%
multisigSM/3             3.658 ms	3.700 ms    +1.1%
multisigSM/4             3.693 ms	3.728 ms    +0.9%
multisigSM/5             4.300 ms	4.343 ms    +1.0%
multisigSM/6             3.607 ms	3.652 ms    +1.2%
multisigSM/7             3.617 ms	3.661 ms    +1.2%
multisigSM/8             3.659 ms	3.697 ms    +1.0%
multisigSM/9             3.697 ms	3.735 ms    +1.0%
multisigSM/10            4.304 ms	4.345 ms    +1.0%
vesting/1                3.162 ms	3.175 ms    +0.4%
vesting/2                2.806 ms	2.826 ms    +0.7%
vesting/3                3.162 ms	3.173 ms    +0.3%
marlowe/trustfund/1      9.995 ms	10.01 ms    +0.2%
marlowe/trustfund/2      7.909 ms	7.934 ms    +0.3%
marlowe/zerocoupon/1     10.15 ms	10.17 ms    +0.2%
marlowe/zerocoupon/2     7.236 ms	7.265 ms    +0.4%
```

This suggests that while this PR does make things a bit worse, most of the slowdown in fact comes from the recent `master` rather than the PR itself. `multisigSM` seems to be affected the most. Is there something specific about it?

@kwxm could you confirm that the previous commit in this branch (i.e. old `master`) is indeed faster than the current `master`? I might have screwed something up... BTW, could you also add to your `bench-compare` a feature that registers the commit of the branch being benchmarked?

**Comment by effectfully** [comment]
And here's the diff between the previous commit in this PR and the last one when it comes to the recursive part of the machine (which is where a slowdown is supposed to be, although I also checked `runCek` and it's basically the same in both the versions): [diff.txt](https://github.com/input-output-hk/plutus/files/6223772/diff.txt)

They are freaking the same modulo names of generated variables and the fact the slow version does not pass the redundant dictionary around and also has jumps to join point instead of recursive calls. I.e. it has be fast and not slow, but for some reason it slow.

**Comment by kwxm** [comment]
> BTW, could you also add to your bench-compare a feature that registers the commit of the branch being benchmarked?

I'm not sure how I could do that.  All it's doing is collecting some data from a couple of text files: it doesn't know where that data came from.  I know what you mean though. I find myself looking at these things and wondering what I was comparing, and worrying that I might have got the comparison the wrong way round.

It'd be nice to have a command that benchmarks two different branches and gives you a report comparing them, but that could take a while and it'd be easy to get carried away adding features..

**Comment by effectfully** [comment]
> This suggests that while this PR does make things a bit worse, most of the slowdown in fact comes from the recent master 

Lol, I'm an idiot. The above comparison (the first table) is between the previous commit in the branch of this PR and the last one. `master` is irrelevant. So somehow the PR makes things worse despite producing Core that looks better, period.

But I *also* compared the previous commit and the current master (the second table) and the current master is slower. So it does seem that something happened in the interim that either slowed down the machine (unlikely, but possible, maybe an updated dependency) or made compilation produce less efficient program (dunno how likely that is).

I'm closing the PR, 'cause I've lost a lot of time here and I don't see what else I could do.

**Pull request closed without merging**